### PR TITLE
[exporter/elasticsearch] support sending otlp tracing to Elasticsearch

### LINKED ISSUE #12069: [exporter/elasticsearch] support sending otlp tracing to Elasticsearch
**Is your feature request related to a problem? Please describe.**
elasticsearch is a commonly used text retrieval engine, so many users will choose elasticsearch as storage for their tracing systems. So, elasticsearch exporter's ability to support traces exports will be more convenient.

### PR DESCRIPTION
**Pull request by JaredTan95**
Signed-off-by: jian.tan <jian.tan@daocloud.io>

support and close https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/12069
**Description:** support sending otlp tracing to Elasticsearch

image

### CODE REVIEW

**Comment by Frapschen** [comment]
cool

**Comment by dmitryax** [comment]
@JaredTan95 please fix liter/test failures

**Comment by JaredTan95** [comment]
> @JaredTan95 please fix linter/test failures

done :-)

**Comment by JaredTan95** [comment]
@codeboten hi, Do you have time to help review the code? :-P

File: exporter/elasticsearchexporter/config.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/config.go:53]: Please keep this option supported but deprecated. It should throw a warning if used

Comment (JaredTan95) [reply exporter/elasticsearchexporter/config.go:53]: got it.

**Comment by JaredTan95** [comment]
BTW, In the future, whether we need to add the [sending_queue](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/exporterhelper)  logic to ensure that the data lost?

File: exporter/elasticsearchexporter/factory.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/factory.go:49]: Why the default value is changed?

File: exporter/elasticsearchexporter/config.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/config.go:226]: Weird that we have default index value `logs-generic-default` that is never being used because the field is required. Do you mind submitting an issue to address that separately. We need to either remove the default value or make the config option not required

File: exporter/elasticsearchexporter/factory.go
Comment (JaredTan95) [reply exporter/elasticsearchexporter/factory.go:49]: mistake chenges

File: exporter/elasticsearchexporter/config.go
Comment (JaredTan95) [reply exporter/elasticsearchexporter/config.go:226]: The logic here was not clear before to me, but if we had the default value by `createDefaultConfig` method, should be no need to detect it here.

Comment (JaredTan95) [reply exporter/elasticsearchexporter/config.go:226]: I think we can remove this logical.

File: exporter/elasticsearchexporter/README.md
Comment (dmitryax) [inline exporter/elasticsearchexporter/README.md:27]: default value is missed here

**Comment by dmitryax** [comment]
cc @urso @faec @blakerouse as code owners, PTAL

File: exporter/elasticsearchexporter/testdata/config-use-deprecated-index_option.yaml
Comment (dmitryax) [inline exporter/elasticsearchexporter/testdata/config-use-deprecated-index_option.yaml:32]:
```suggestion
```

File: exporter/elasticsearchexporter/factory.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/factory.go:86]: Please add a warning message if the deprecated field is being used

Comment (JaredTan95) [reply exporter/elasticsearchexporter/factory.go:86]: make sense

File: exporter/elasticsearchexporter/factory.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/factory.go:50]: I would make `Index=""` by default

File: exporter/elasticsearchexporter/logs_exporter.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/logs_exporter.go:75]: This doesn't seem to be a backward compatible solution. We need to use `cfg.Index` if it's provided and `cfg.LogsIndex` is not.

Comment (JaredTan95) [reply exporter/elasticsearchexporter/logs_exporter.go:77]: thx, I updated this logical.

File: exporter/elasticsearchexporter/factory.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/factory.go:75]:
```suggestion
	if cfg.(*Config).Index != "" {
```

File: exporter/elasticsearchexporter/logs_exporter.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/logs_exporter.go:71]: Let's not mutate the original config. Introduce a new variable for the index instead

File: exporter/elasticsearchexporter/trace_exporter.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/trace_exporter.go:108]:
```suggestion
func (e *elasticsearchTracesExporter) pushTraceRecord(ctx context.Context, resource pcommon.Resource, span ptrace.Span) error {
	document, err := e.model.encodeSpan(resource, span)
```

File: exporter/elasticsearchexporter/trace_exporter.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/trace_exporter.go:115]: This function seems to be pretty much the same as `pushRecord`. Can we avoid the duplication?

Comment (JaredTan95) [reply exporter/elasticsearchexporter/trace_exporter.go]: make sense

**Comment by JaredTan95** [comment]
@jpkrohling @codeboten @dmitryax hi, this PR ready to review :)

File: exporter/elasticsearchexporter/logs_exporter.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/logs_exporter.go:76]: Looks like `cfg.LogsIndex` is not being used at all

Comment (JaredTan95) [reply exporter/elasticsearchexporter/logs_exporter.go:76]: ahh, Important logical missed, I fixed it and add some UT to cover it.

File: exporter/elasticsearchexporter/logs_exporter.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/logs_exporter.go:73]: I don't get why `& cfg.Index != defaultLogsIndex` part is needed?

Maybe just 
```suggestion
	indexStr := cfg.LogsIndex
	if cfg.Index != "" {
		indexStr = cfg.Index
	}
```

File: exporter/elasticsearchexporter/trace_exporter.go
Comment (dmitryax) [inline exporter/elasticsearchexporter/trace_exporter.go:92]: Not for this PR, but would be nice to have an issue if you agree:

Currently we send data with one request per trace / log record. I see couple issues with that:
1. I expect it to be not very performant.
2. Metrics reported about failed pushes are based on `error` returned by this function. It means that if even just one request failed, the whole `resourceSpans` batch is marked as failed to be sent.

I would recommend looking into an option for making one requests to send the whole batch of logs or traces

Comment (JaredTan95) [reply exporter/elasticsearchexporter/trace_exporter.go:92]: got it, I plan to do this after PR.

**Review by dmitryax — APPROVED** [review]
LGTM. Pinging code owners as well @urso @faec @blakerouse

**Comment by dmitryax** [comment]
I would be good to have at least one approval from code owners before merging it

**Comment by jpkrohling** [comment]
It would be good to have a review from a current code owner as well: cc
 @urso @faec @blakerouse

**Comment by jpkrohling** [comment]
Sorry for closing/opening, I clicked the wrong button.

**Comment by JaredTan95** [comment]
hi, @urso @faec @blakerouse do you have time to help review this feature PR?

**Comment by ramdaspotale** [comment]
Hi, @JaredTan95 

I am eagerly waiting for this PR to be completed as soon as possible, as this functionality is much needed in my case. can you please review failed required checks and help to merge this?

**Review by dmitryax — APPROVED** [review]

**Pull request merged**
